libjlm/include/jlm/ir/operators/operators.hpp

```diff
@@ -1100,6 +1100,81 @@ create_fpext_tac(const variable * operand, jlm::variable * result)
 	return tac::create(op, {operand}, {result});
 }
 
+/* fpneg operator */
+
+class fpneg_op final : public jive::unary_op {
+public:
+	virtual
+	~fpneg_op() noexcept;
+
+	inline
+	fpneg_op(const jlm::fpsize & srcsize, const jlm::fpsize & dstsize)
+	: unary_op(fptype(srcsize), fptype(dstsize))
+	{
+		if (srcsize == fpsize::flt && dstsize == fpsize::half)
+			throw jlm::error("destination type size must be bigger than source type size.");
+	}
+
+	inline
+	fpneg_op(
+		std::unique_ptr<jive::type> srctype,
+		std::unique_ptr<jive::type> dsttype)
+	: unary_op(*srctype, *dsttype)
+	{
+		auto st = dynamic_cast<const jlm::fptype*>(srctype.get());
+		if (!st) throw jlm::error("expected floating point type.");
+
+		auto dt = dynamic_cast<const jlm::fptype*>(dsttype.get());
+		if (!dt) throw jlm::error("expected floating point type.");
+
+		if (st->size() == fpsize::flt && dt->size() == fpsize::half)
+			throw jlm::error("destination type size must be bigger than source type size.");
+	}
+
+	virtual bool
+	operator==(const operation & other) const noexcept override;
+
+	virtual std::string
+	debug_string() const override;
+
+	virtual std::unique_ptr<jive::operation>
+	copy() const override;
+
+	jive_unop_reduction_path_t
+	can_reduce_operand(
+		const jive::output * output) const noexcept override;
+
+	jive::output *
+	reduce_operand(
+		jive_unop_reduction_path_t path,
+		jive::output * output) const override;
+
+	inline const jlm::fpsize &
+	srcsize() const noexcept
```

Comment:
There is only a size function necessary, and not a srcsize and dstsize, as the input and output operand have the same size: const jlm::fpsize & size() const noexcept {...}